src/lib.rs

```diff
@@ -1,91 +1,64 @@
-use cfgrammar::RIdx;
-use lrlex::{lrlex_mod, DefaultLexeme, DefaultLexerTypes, LRNonStreamingLexerDef};
-use lrpar::{lrpar_mod, Lexeme, Node};
+use lrlex::{lrlex_mod, DefaultLexerTypes};
+use lrpar::{lrpar_mod, NonStreamingLexer};
 
+// Using `lrlex_mod!` brings the lexer for `calc.l` into scope. By default the module name will be
+// `calc_l` (i.e. the file name, minus any extensions, with a suffix of `_l`).
 lrlex_mod!("calc.l");
+// Using `lrpar_mod!` brings the parser for `calc.y` into scope. By default the module name will be
+// `calc_y` (i.e. the file name, minus any extensions, with a suffix of `_y`).
 lrpar_mod!("calc.y");
 
-pub fn eval(
-    lexerdef: &LRNonStreamingLexerDef<DefaultLexerTypes>,
-    input: &str,
-) -> Result<String, String> {
-    let lexer = lexerdef.lexer(input);
-    let (res, errs) = calc_y::parse(&lexer);
-    for e in errs {
-        println!("{}", e.pp(&lexer, &calc_y::token_epp));
-    }
-    match res {
-        Some(x) => {
-            return Ok(format!("{}", Eval::new(input).eval(&x)));
+use calc_y::Expr;
+
+#[derive(Debug)]
+pub struct Calc {}
+
+impl Calc {
+    pub fn from_str(input: &str) -> Result<u64, &'static str> {
+        let exp = self::parse(input);
+        if let Ok(exp) = exp {
+            let lexerdef = calc_l::lexerdef();
+            let lexer = lexerdef.lexer(input);
+            return eval(&lexer, exp);
         }
-        _ => Err(String::from("Unable to evaluate expression.")),
+        return Err("Unable to evaluate expression.");
     }
 }
 
-struct Eval<'a> {
-    s: &'a str,
-}
-
-// stolen from: https://github.dev/softdevteam/grmtools
-impl<'a> Eval<'a> {
-    fn new(s: &'a str) -> Self {
-        return Eval { s };
+pub fn parse(input: &str) -> Result<Expr, &str> {
+    let lexerdef = calc_l::lexerdef();
+    let lexer = lexerdef.lexer(input);
+    let (res, errs) = calc_y::parse(&lexer);
+    for e in errs {
+        println!("{}", e.pp(&lexer, &calc_y::token_epp));
     }
 
-    fn eval_r_exp(&self, nodes: &Vec<Node<DefaultLexeme, u32>>) -> i64 {
-        if nodes.len() == 1 {
-            return self.eval(&nodes[0]);
-        } else {
-            debug_assert_eq!(nodes.len(), 3);
-            return self.eval(&nodes[0]) + self.eval(&nodes[2]);
-        }
+    if let Some(Ok(r)) = res {
+        return Ok(r);
     }
+    return Err("Something went wrong!");
```

Comment:
You can be more specific here. How about including the stringified underlying error?

Reply:
Added

Reply:
This line was removed. All errors messages are accumulated here: https://github.com/softdevteam/pavel.yaiwr/blob/39bade38487260a6eed5a5d96993466b485ba363/src/lib.rs#L19